homeassistant/components/roborock/sensor.py

```diff
@@ -210,11 +269,23 @@ async def async_setup_entry(
         )
         for device_id, coordinator in coordinators.items()
         for description in SENSOR_DESCRIPTIONS
+        if isinstance(coordinator, RoborockDataUpdateCoordinator)
```

Comment:
Instead of all these isinstance checks can we not have two dictionaries or a custom data class

Reply:
I'm not following how that helps us here. I need to know what type of coordinator it is for each one as there is a different support check i do for both. Either  if description.value_fn(coordinator.roborock_device_info.props) is not None
    ) or if description.data_protocol in coordinator.data

Reply:
I meant something like:

```diff
diff --git a/homeassistant/components/roborock/sensor.py b/homeassistant/components/roborock/sensor.py
index bdb5ae586e..ba478416e0 100644
--- a/homeassistant/components/roborock/sensor.py
+++ b/homeassistant/components/roborock/sensor.py
@@ -252,35 +252,37 @@ A01_SENSOR_DESCRIPTIONS: list[RoborockSensorDescriptionA01] = [
 ]
 
 
+@dataclass
+class RoborockCoordinators:
+    """Roborock coordinators type."""
+
+    v1: list[RoborockDataUpdateCoordinator]
+    a01: list[RoborockDataUpdateCoordinatorA01]
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up the Roborock vacuum sensors."""
-    coordinators: dict[
-        str, RoborockDataUpdateCoordinator | RoborockDataUpdateCoordinatorA01
-    ] = hass.data[DOMAIN][config_entry.entry_id]
+    coordinators: RoborockCoordinators = hass.data[DOMAIN][config_entry.entry_id]
     async_add_entities(
         RoborockSensorEntity(
-            f"{description.key}_{slugify(device_id)}",
             coordinator,
             description,
         )
-        for device_id, coordinator in coordinators.items()
+        for coordinator in coordinators.v1
         for description in SENSOR_DESCRIPTIONS
-        if isinstance(coordinator, RoborockDataUpdateCoordinator)
         if description.value_fn(coordinator.roborock_device_info.props) is not None
     )
     async_add_entities(
         RoborockSensorEntityA01(
-            f"{description.key}_{slugify(device_id)}",
             coordinator,
             description,
         )
-        for device_id, coordinator in coordinators.items()
+        for coordinator in coordinators.a01
         for description in A01_SENSOR_DESCRIPTIONS
-        if isinstance(coordinator, RoborockDataUpdateCoordinatorA01)
         if description.data_protocol in coordinator.data
     )
 
@@ -292,13 +294,16 @@ class RoborockSensorEntity(RoborockCoordinatedEntityV1, SensorEntity):
 
     def __init__(
         self,
-        unique_id: str,
         coordinator: RoborockDataUpdateCoordinator,
         description: RoborockSensorDescription,
     ) -> None:
         """Initialize the entity."""
         self.entity_description = description
-        super().__init__(unique_id, coordinator, description.protocol_listener)
+        super().__init__(
+            f"{description.key}_{slugify(coordinator.roborock_device_info.device.duid)}",
+            coordinator,
+            description.protocol_listener,
+        )
 
     @property
     def native_value(self) -> StateType | datetime.datetime:
@@ -315,13 +320,15 @@ class RoborockSensorEntityA01(RoborockCoordinatedEntityA01, SensorEntity):
 
     def __init__(
         self,
-        unique_id: str,
         coordinator: RoborockDataUpdateCoordinatorA01,
         description: RoborockSensorDescriptionA01,
     ) -> None:
         """Initialize the entity."""
         self.entity_description = description
-        super().__init__(unique_id, coordinator)
+        super().__init__(
+            f"{description.key}_{slugify(coordinator.roborock_device_info.device.duid)}",
+            coordinator,
+        )
 
     @property
     def native_value(self) -> StateType:
```

The definition of `RoborockCoordinators` should move to the init file, but I added here to give you can example
Probably the postfix of the unique_id (`slugify(coordinator.roborock_device_info.device.duid)`) could be moved to the base entity

Reply:
Ah okay - I like this, thanks for the suggestion

Reply:
Changes made!